src/library_async.js

```diff
@@ -828,6 +889,137 @@ mergeInto(LibraryManager.library, {
     });
   },
 
+  $Fibers: {
+    id: 0,  // last allocated fiber ID
```

Comment:
I know you've said you prefer the current approach,  but what do you think about prototyping a version of this change with the following 3 changes:

1.  Add `current` here which stores the current id (initialized to 0)
2.  Rename `create_from_current` to just `get_current` to just return `Fibres.current`
3.  Initialize Fibers[0] to be the empty object {}
4.  emscripten_fiber_swap just takes a single target fibre id argument.

I think it would make the API nicer.    I think (4) is the only thing that would be visible to the API user.   I'm trying to imagine to downside to doing (4) but I can't think of one.   Is there use case for swap where the old fiber is not the current fibre?

Reply:
Okay. I've pushed the prototype into a new `fibers-alt` branch on my fork: https://github.com/Akaricchi/emscripten/commit/641a257572abe91db21d926147dd88b14507104a
I've also pushed an `emscripten-fibers-alt` branch at https://github.com/taisei-project/koishi/commit/15017b49328f043880c9c37d7af5cabdf6eebae9 for testing.

I deviated from your proposed changes a little, because:
1. `0` is supposed to mean NULL/invalid fiber
2. Initializing the "main" fiber to an empty object does not work, it needs `stackBase` and `stackMax` set up, *unless* `emscripten_fiber_swap` is also modified to set them up on every swap, which is wasteful.

For the above reasons, I had to make `emscripten_fiber_get_current` check whether the "current" fiber is valid, and if not, create a partial fiber (just as `emscripten_fiber_create_from_current_context` would). `emscripten_fiber_swap` now calls `emscripten_fiber_get_current` to retrieve the "old fiber" handle.

The result is slightly increased code size and apparent ~5% performance loss on koishi's microbenchmark. It's not disasterous, but I still don't think it's worth the slight API convenience. Especially considering that these fibers are already hundreds of times slower than native equivalents (which are often implemented in a few lines of assembly and expected to be *very* fast).

I also still have no idea how this is going to work with threads. Does each thread have its own JS VM/global namespace?

I'd also like to note that even the Win32 Fibers API that you've brought up here earlier does not let you use any of the fiber functions without explicitly creating a "main" fiber for the thread first. [They just did it in a weird-ass way, much like everything else Microsoft does.](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-convertthreadtofiber) Here's [libkoishi's win32fiber backend](https://github.com/taisei-project/koishi/blob/master/src/win32fiber/win32fiber.c) for a basic example.

TL;DR I still don't like this, but it's not a hill to die on; I'll implement this if I have to.

Reply:
> Is there use case for swap where the old fiber is not the current fibre?

Probably not. The only point of that argument is specifically to avoid tracking the current fiber in runtime.

Reply:
Thanks!  I appreciate you entertaining my idea.    I'll take a look the alt version tomorrow.   Do you have any idea why it would be 5% slowdown?    It seems like the swap operation should take exactly the same amount of time in either case.  maybe I'm missing something? 

In terms of threads, its little complicated but currently on the web each thread has its own JS context, yes.   The only thing they share is the linear memory.    I hoping limiting fibres to running only on threads that created them is reasonable limitation.   It certainly seems reasonable to me, but maybe I'm not seeing some important use case?  

In a push, we send messages between threads using postMessage() or serialize data to shared memory and using that communicate.  But that certainly seems out-of-scope for the initial implementation.

Reply:
> Do you have any idea why it would be 5% slowdown? It seems like the swap operation should take exactly the same amount of time in either case. maybe I'm missing something?

It can't be exactly the same because it needs to also call into `emscripten_fiber_get_current`, which might need to create a fiber for the first time, so there's always at least an additional access to the `Fibers` object and a comparison. The 5% is a very crude and probably inaccurate figure, but there's objectively *some* bit of extra work done on every swap now, however small.

> I hoping limiting fibres to running only on threads that created them is reasonable limitation. It certainly seems reasonable to me, but maybe I'm not seeing some important use case?

It's probably an okay limitation for most use cases, including mine, but I'm pretty sure that some game engines have an M:N concurrency model where a coroutine/fiber may be resumed on any of the threads. I don't have much experience with that kind of code so I can't give you a useful example. And this limitation seems difficult to overcome with the tools we're working with, if at all possible.

While non-critical, it may be annoying and confusing that coroutine handles are also scoped per-thread now; i.e. handle 1 on thread 1 is not the same as handle 1 on thread 2, even though they appear bitwise identical on the WASM side. This may be solved by moving the handle counter into a global in the shared memory, but we'll need to ensure atomic access somehow.

Reply:
> It can't be exactly the same because it needs to also call into `emscripten_fiber_get_current`, which might need to create a fiber for the first time, so there's always at least an additional access to the `Fibers` object and a comparison. The 5% is a very crude and probably inaccurate figure, but there's objectively _some_ bit of extra work done on every swap now, however small.

Right, but at the same time there is less work being done on wasm side (no need to keep track or current fibre in all place in C++) and one fewer argument being passed between the wasm-JS boundary.

I assumed that top of swap would just have something like `if (!current) current = get_current;` at the top?  Which would be just a single rarly taken branch on the JS side, which I assumed would be a net win.  But I still haven't look at the code yet..

Reply:
> Right, but at the same time there is less work being done on wasm side (no need to keep track or current fibre in all place in C++)

This is something I've been saying all this time: **in a realistic use case, the wasm side will track the current fiber anyway**. Sure, the trivial test case I've included does not need to track the current fiber anymore (in fact it never did, I just hardcoded the "old fiber" arguments). But something like libkoishi (a pretty basic implementation of asymmetrical coroutines) needs to know the current fiber *and a bunch of state associated with it*, such as which coroutine it was entered from so that it knows where to yield. So it ends up tracking its own coroutine structure, which already contains all the necessary information, including the fiber handle.